src/planner/vectorfield/detail/VectorFieldIntegrator.cpp

```diff
@@ -0,0 +1,172 @@
+{
+  using dart::common::make_unique;
+
+  std::size_t dimension = stateSpace->getDimension();
+
+  // TODO: const cast will be removed after the spline constructor updated.
+  auto nonConstStateSpace
+      = std::const_pointer_cast<aikido::statespace::StateSpace>(stateSpace);
+  auto outputTrajectory
+      = make_unique<aikido::trajectory::Spline>(nonConstStateSpace);
+
+  using CubicSplineProblem = aikido::common::
+      SplineProblem<double, int, 2, Eigen::Dynamic, Eigen::Dynamic>;
+
+  const Eigen::VectorXd zeroPosition = Eigen::VectorXd::Zero(dimension);
+  auto currState = stateSpace->createState();
+  for (std::size_t iknot = 0; iknot < knots.size() - 1; ++iknot)
+  {
+    const double segmentDuration = knots[iknot + 1].mT - knots[iknot].mT;
+    Eigen::VectorXd currentPosition = knots[iknot].mPositions;
+    Eigen::VectorXd nextPosition = knots[iknot + 1].mPositions;
+
+    CubicSplineProblem problem(
+        Eigen::Vector2d{0., segmentDuration}, 2, dimension);
+    problem.addConstantConstraint(0, 0, zeroPosition);
+    problem.addConstantConstraint(1, 0, nextPosition - currentPosition);
+    const auto solution = problem.fit();
+    const auto coefficients = solution.getCoefficients().front();
+
+    stateSpace->expMap(currentPosition, currState);
+    outputTrajectory->addSegment(coefficients, segmentDuration, currState);
+  }
+  return outputTrajectory;
+}
+
+//==============================================================================
+VectorFieldIntegrator::VectorFieldIntegrator(
+    const VectorField* vectorField,
+    const aikido::constraint::Testable* collisionFreeConstraint,
+    double timelimit,
+    double checkConstraintResolution)
+  : mVectorField(vectorField)
+  , mConstraint(collisionFreeConstraint)
+  , mTimelimit(timelimit)
+  , mConstraintCheckResolution(checkConstraintResolution)
+{
+  mCacheIndex = -1;
+  mIndex = 0;
+
+  mDimension = mVectorField->getStateSpace()->getDimension();
+}
+
+//==============================================================================
+void VectorFieldIntegrator::start()
+{
+  mTimer.start();
+  mKnots.clear();
+  mCacheIndex = -1;
+  mIndex = 0;
+}
+
+//==============================================================================
+int VectorFieldIntegrator::getCacheIndex()
+{
+  return mCacheIndex;
+}
+
+//==============================================================================
+std::vector<Knot>& VectorFieldIntegrator::getKnots()
+{
+  return mKnots;
+}
+
+//==============================================================================
+void VectorFieldIntegrator::step(
+    const Eigen::VectorXd& q, Eigen::VectorXd& qd, double /*_t*/)
+{
+  aikido::statespace::StateSpace::State* state
+      = mVectorField->getStateSpace()->allocateState();
+  mVectorField->getStateSpace()->expMap(q, state);
+
+  // compute joint velocities
+  bool success = mVectorField->evaluateVelocity(state, qd);
+  mVectorField->getStateSpace()->freeState(state);
```

Comment:
You can eliminate this manual `freeState()` call if you call `createState()` instead of `allocateState()`. That function returns a `ScopedState`, which is a RAII wrapper around `State`.

This line also disappears if we store the state as a member variable. See my suggestion above.